src/backend/lib/utils.js

```diff
@@ -0,0 +1,299 @@
+// @ts-check
+const semver = require("semver");
+const hasha = require("hasha");
+const path = require("path");
+const fs = require("fs");
+const stream = require("stream");
+const { promisify } = require("util");
+const { once } = require("events");
+const pDefer = require("p-defer");
+const copyError = require("utils-copy-error");
+const ApkParser = require("app-info-parser/src/apk");
+const ZipFs = require("@gmaclennan/zip-fs");
+
+/** @typedef {import('./types').InstallerInt} InstallerInt */
+/** @typedef {import('./types').InstallerExt} InstallerExt */
+
+module.exports = {
+  installerCompare,
+  isUpgradeCandidate,
+  getBestUpgradeCandidate,
+  getInstallerInfo,
+  beforeAfterStream,
+  stringifyInstaller,
+};
+
+/**
+ * Compare two installer options, returns -1 if a is less than b, 0 if same, 1
+ * if a is greater than b
+ *
+ * @param {InstallerInt | InstallerExt} a
+ * @param {InstallerInt | InstallerExt} b
+ * @returns {-1 | 0 | 1}
+ */
+function installerCompare(a, b) {
+  const aIsValidSemver = semver.valid(a.versionName);
+  const bIsValidSemver = semver.valid(b.versionName);
+  if (!aIsValidSemver && !bIsValidSemver) return 0;
+  if (!aIsValidSemver) return -1;
+  if (!bIsValidSemver) return 1;
+  // Important to validate semver before here, otherwise this will throw
+  const semverCompare = semver.compare(a.versionName, b.versionName);
+  if (semverCompare !== 0) return semverCompare;
+  if (a.versionCode < b.versionCode) return -1;
+  if (a.versionCode > b.versionCode) return 1;
+  return 0;
+}
+
+/**
+ * Check if an installer is compatible with the given device. Requirements:
+ *
+ * 1. Must support arch of device
+ * 2. Must support SDK of device
+ * 3. Version code must be greater or equal to currently installed app
```

Comment:
You are correct, it should not be equal. Android will allow you to install an app with an equal version code, but if we allow that it will allow an inifinite loop of updates (imagine an APK with the same version name, e.g. `1.0.0` as the current app, and the same version code, but a different hash. It would be considered an update, but then in the other direction is also considered an update). We use build number for version code, so for Mapeo `versionCode` = build number. Build number is guaranteed to increment for every build of an APK we create on our CI.

Reply:
Luandro is fixing this in https://github.com/digidem/mapeo-mobile/pull/596

Reply:
:+1: